source/Halibut/ServiceModel/HalibutProxyWithAsync.cs

```diff
@@ -0,0 +1,172 @@
+
+namespace Halibut.ServiceModel
+{
+    public class HalibutProxyWithAsync : DispatchProxyAsync
+    {
+        Func<RequestMessage, MethodInfo, CancellationToken, Task<ResponseMessage>> messageRouter;
+        Type contractType;
+        ServiceEndPoint endPoint;
+        long callId;
+        bool configured;
+        CancellationToken globalCancellationToken;
+        ILog logger;
+
+        public void Configure(Func<RequestMessage, MethodInfo, CancellationToken, Task<ResponseMessage>> messageRouter, Type contractType, ServiceEndPoint endPoint, ILog logger, CancellationToken cancellationToken)
+        {
+            this.messageRouter = messageRouter;
+            this.contractType = contractType;
+            this.endPoint = endPoint;
+            this.globalCancellationToken = cancellationToken;
+            this.configured = true;
+            this.logger = logger;
+        }
+
+        public override object Invoke(MethodInfo targetMethod, object[] args)
+        {
+            throw new NotImplementedException();
+        }
+
+        public override async Task InvokeAsync(MethodInfo asyncMethod, object[] args)
+        {
+            var (serviceMethod, result) = await MakeRpcCall(asyncMethod, args);
+        }
+
+        public override async Task<T> InvokeAsyncT<T>(MethodInfo asyncMethod, object[] args)
+        {
+            var (serviceMethod, result) = await MakeRpcCall(asyncMethod, args);
+
+            var returnType = serviceMethod.ReturnType;
+            if (result != null && returnType != typeof(void) && !returnType.IsInstanceOfType(result))
+            {
+                result = (T) Convert.ChangeType(result, returnType);
+            }
+
+            return (T) result;
+        }
+
+        async Task<(MethodInfo, object)> MakeRpcCall(MethodInfo asyncMethod, object[] args)
+        {
+            var serviceMethod = AsyncCompatibilityHelper.FindMatchingSyncMethod(asyncMethod, contractType);
+
+            if (!configured)
+                throw new Exception("Proxy not configured");
+
+            var trimmedArgsAndHalibutProxyRequestOptions = TrimOffHalibutProxyRequestOptions(args);
+            args = trimmedArgsAndHalibutProxyRequestOptions.args;
+            var halibutProxyRequestOptions = trimmedArgsAndHalibutProxyRequestOptions.halibutProxyRequestOptions;
+
+            var request = CreateRequest(serviceMethod, args);
+
+            var response = await messageRouter(request, serviceMethod, ConnectingCancellationToken(halibutProxyRequestOptions));
+
+            EnsureNotError(response);
+            
+            return (serviceMethod, response.Result);
+        }
+
+        RequestMessage CreateRequest(MethodInfo targetMethod, object[] args)
+        {
+            var activityId = Guid.NewGuid();
+
+            var request = new RequestMessage
+            {
+                Id = contractType.Name + "::" + targetMethod.Name + "[" + Interlocked.Increment(ref callId) + "] / " + activityId,
```

Comment:
Would it be useful to hint that this was done via an async call somehow? The activity id gets used in the LogMessage so may be useful for diagnostic purposes

Reply:
good idea, done!